src/main/java/com/redhat/devtools/lsp4ij/features/documentSymbol/LSPBreadcrumbsProvider.java

```diff
@@ -0,0 +1,148 @@
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *
+ * Contributors:
+ *     Red Hat Inc. - initial API and implementation
+ *******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.documentSymbol;
+
+import com.intellij.ide.structureView.StructureViewTreeElement;
+import com.intellij.lang.Language;
+import com.intellij.lang.LanguageUtil;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Conditions;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.ui.breadcrumbs.BreadcrumbsProvider;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.LanguageServiceAccessor;
+import com.redhat.devtools.lsp4ij.features.documentSymbol.LSPDocumentSymbolStructureViewModel.LSPDocumentSymbolViewElement;
+import com.redhat.devtools.lsp4ij.features.documentSymbol.LSPDocumentSymbolStructureViewModel.LSPFileStructureViewElement;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.Icon;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * LSP document symbol-based breadcrumbs provider.
+ */
+public class LSPBreadcrumbsProvider implements BreadcrumbsProvider {
+
+    @Override
+    public Language[] getLanguages() {
+        // Register for all languages and filter in isSupported()
+        List<Language> languages = LanguageUtil.getLanguages(Conditions.alwaysTrue());
+        return languages.toArray(Language.EMPTY_ARRAY);
+    }
+
+    private boolean isSupported(@NotNull PsiElement element) {
+        Project project = element.getProject();
+        PsiFile file = element.getContainingFile();
+        VirtualFile virtualFile = file.getVirtualFile();
+        return LanguageServiceAccessor.getInstance(project).hasAny(
+                virtualFile,
+                // The breadcrumbs feature must be enabled and supported for the file's language server
+                ls -> ls.getClientFeatures().getBreadcrumbsFeature().isEnabled(file) &&
+                        ls.getClientFeatures().getBreadcrumbsFeature().isSupported(file)
+        );
+    }
+
+    @Override
+    public boolean acceptElement(@NotNull PsiElement element) {
+        // NOTE: This is only used to determine which already-vetted elements should be included in the breadcrumb trail
+        return element instanceof DocumentSymbolData;
+    }
+
+    @Override
+    @Nullable
+    public PsiElement getParent(@NotNull PsiElement element) {
+        if (isSupported(element)) {
```

Comment:
I wonder if we should do that for each element, since document symbol belongs to the same file?

Reply:
Assuming you mean using `CachedValuesManager` for caching, we'd need to add a modification tracker for language server configuration that causes a read-through if/when language server configs are changed. Otherwise cached values could become stale.

I'd probably lump this under the "potential premature optimization" category and would consider it if we observe performance issues around this check. I've not seen anything that concerns me even in large files.

Reply:
I agree, it was just a question. I think we have no performance problem with this isSupported call

Reply:
It does raise an interesting point, though. There might be opportunities (and need) for caching where cached elements should be evicted automatically when the associated language server definition config changes. If/when that's needed for sure, it would make sense to add a `ModificationTracker` implementation to `LanguageServerDefinition` that can be used as a cache dependency and is automatically incremented when that language server definition changes in any manner that should result in cache eviction.

Something to consider if/when it's needed...

Reply:
I think more and more that we need to cache for all lsp features the isSupported in the psifile for each associated ls. See https://github.com/redhat-developer/lsp4ij/issues/7